feat: add C implementation for `math/base/special/sici`

navyansh007 opened the pull request:
resolves #7524

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

navyansh007 committed:
feat: added C implementation for math/base/special/sici

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

navyansh007 committed:
style: handled the invalid year issue in some files

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/math/base/special/sici/src/main.c:
```diff
@@ -0,0 +1,385 @@
+			}
+			return;
+		}
+		*si = STDLIB_CONSTANT_FLOAT64_HALF_PI - ( stdlib_base_cos( ax ) / ax );
```
@anandkaranubc Here and below would be good opportunities to use `sincos`. I see that this also applies to the JS implementation.

navyansh007 replied:
Hey @kgryte, thanks for pointing out this one. I wasn't aware that there is a function called `sincos` that actually calculates both `sin` and `cos` values simultaneously, removing the dependency of calling separate `sin` and `cos` functions. Hence, I have now incorporated that in my `sici` function's C implementation.

navyansh007 committed:
refactor: incorporated sincos instead of separate sin and cos functions

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sici/manifest.json:
Hey there @navyansh007, you have used `sincos` in the C implementation in `main.c`, but that dependency is not listed in this file. You'll need to add that here.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sici/manifest.json:
```diff
@@ -0,0 +1,93 @@
+        "@stdlib/math/base/special/sin",
+        "@stdlib/constants/float64/half-pi",
+        "@stdlib/constants/float64/eulergamma",
+        "@stdlib/constants/float64/ninf"
```
```suggestion
        "@stdlib/constants/float64/ninf",
        "@stdlib/math/base/special/sincos"
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sici/manifest.json:
```diff
@@ -0,0 +1,93 @@
+        "@stdlib/math/base/special/sin",
+        "@stdlib/constants/float64/half-pi",
+        "@stdlib/constants/float64/eulergamma",
+        "@stdlib/constants/float64/ninf"
```
```suggestion
        "@stdlib/constants/float64/ninf",
        "@stdlib/math/base/special/sincos"
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sici/manifest.json:
```diff
@@ -0,0 +1,93 @@
+        "@stdlib/math/base/special/sin",
+        "@stdlib/constants/float64/half-pi",
+        "@stdlib/constants/float64/eulergamma",
+        "@stdlib/constants/float64/ninf"
```
```suggestion
        "@stdlib/constants/float64/ninf",
        "@stdlib/math/base/special/sincos"
```

gunjjoshi committed:
Apply suggestions from code review

Signed-off-by: Gunj Joshi <gunjjoshi8372@gmail.com>

gunjjoshi committed:
add new directories

Signed-off-by: Gunj Joshi <gunjjoshi8372@gmail.com>

gunjjoshi committed:
Update package.json

Signed-off-by: Gunj Joshi <gunjjoshi8372@gmail.com>

gunjjoshi commented:
@navyansh007 You'll need to update `README.md`, which is currently missing from this PR. For reference, you can have a look at [this](https://github.com/stdlib-js/stdlib/pull/2751/files#diff-6d2cbfb65c3d4502e6ba35d8272105c50cf8c87e5ff0a444a1eb0dc1b9db0a8e) as an example.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sici/src/main.c:
```diff
@@ -0,0 +1,385 @@
+	}
+
+	*si = s;
+	*ci = STDLIB_CONSTANT_FLOAT64_EULERGAMMA + log( ax ) + c;
```
@navyansh007 You can use [`@stdlib/math/base/special/ln`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/ln) here, instead of using `log()` from `math.h`.

navyansh007 commented:
Thanks for your feedback @gunjjoshi. Going to work on all of those feedbacks.

navyansh007 committed:
refactor: implemented usage of math/base/special/ln

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

navyansh007 committed:
docs: added C interface documentation of sici

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

navyansh007 committed:
Merge branch 'stdlib-js:develop' into sici-c-implementation

navyansh007 commented:
Hey @kgryte @gunjjoshi 
I have implemented all the changes suggested by @gunjjoshi in the `sici` function's C implementation. Feel free to let me know if any further modifications are required.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sici/manifest.json:
```diff
@@ -0,0 +1,99 @@
+        "@stdlib/math/base/assert/is-infinite",
+        "@stdlib/math/base/assert/is-nan",
+        "@stdlib/math/base/special/cos",
+        "@stdlib/math/base/special/sin",
```
```suggestion
```
We do not need these now, as we're using `stdlib_base_sincos`.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sici/manifest.json:
```diff
@@ -0,0 +1,99 @@
+        "@stdlib/math/base/assert/is-infinite",
+        "@stdlib/math/base/assert/is-nan",
+        "@stdlib/math/base/special/cos",
+        "@stdlib/math/base/special/sin",
```
```suggestion
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sici/manifest.json:
```diff
@@ -0,0 +1,99 @@
+        "@stdlib/math/base/assert/is-infinite",
+        "@stdlib/math/base/assert/is-nan",
+        "@stdlib/math/base/special/cos",
+        "@stdlib/math/base/special/sin",
```
```suggestion
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sici/scripts/evalpoly.js:
```diff
@@ -167,13 +199,16 @@ header += '\n/* This is a generated file. Do not edit directly. */\n';
 		'encoding': 'utf8'
 	};
 
+	// Generate JavaScript files:
```
```suggestion
```

Let's not keep this, as none of the other packages such as [`exp`](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/exp/scripts/evalpoly.js), [`atan`](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/atan/scripts/evalpoly.js), etc. have such comments. The initialization of `opts` and `copts` should be enough to have a visual separation between the generation of JS and C files.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sici/scripts/evalpoly.js:
```diff
@@ -221,6 +256,65 @@ function main() {
 	str = header + compile( GD8 );
 	writeFileSync( fpath, str, opts );
+
+	// Generate C files:
```
```suggestion
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sici/src/main.c:
```diff
@@ -0,0 +1,384 @@
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
```
```suggestion
*
*
* ## Notice
*
* The original C code, long comment, copyright, license, and constants are from [Cephes]{@link http://www.netlib.org/cephes}. The implementation follows the original, but has been modified according to project conventions.
*
* ```text
* Copyright 1984, 1987, 1989 by Stephen L. Moshier
*
* Some software in this archive may be from the book _Methods and Programs for Mathematical Functions_ (Prentice-Hall or Simon & Schuster International, 1989) or from the Cephes Mathematical Library, a commercial product. In either event, it is copyrighted by the author. What you see here may be used freely but it comes with no support or guarantee.
*
* Stephen L. Moshier
* moshier@na-net.ornl.gov
* ```
*/
```

Let's preserve the full header comment from `main.js`.

gunjjoshi commented:
@navyansh007 Would you mind replacing the use of `sin` and `cos` in `main.js` by `sincos` as well, as per https://github.com/stdlib-js/stdlib/pull/8187#discussion_r2399905278?

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sici/src/main.c:
```diff
@@ -0,0 +1,384 @@
+*
+* @param x     input value
+* @param si    output pointer for sine integral
+* @param ci    output pointer for cosine integral
```
```suggestion
* @param si    destination to store the sine integral
* @param ci    destination to store the cosine integral
```

Following what we have in `sincos`: https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/sincos/src/main.c#L97

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sici/src/main.c:
```diff
@@ -0,0 +1,384 @@
+* stdlib_base_sici( 3.0, &si, &ci );
+*/
+void stdlib_base_sici( const double x, double *si, double *ci ) {
+	double sgn;
```
I think we can keep `sgn` as an 8 bit integer, instead of declaring it as double.

gunjjoshi replied:
And that would then require us to make changes at all other places in this function which use `sgn`, such as replacing `sgn = -1.0` by `sgn = -1`. The cephes implementation too, doesn't keep `sgn` as `double`.

navyansh007 committed:
Merge branch 'stdlib-js:develop' into sici-c-implementation

navyansh007 committed:
refactor: removed the usage of sin and cos dependencies from manifest.json

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

navyansh007 committed:
style: removed comments to keep standards at par with other namespace packages

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

navyansh007 committed:
docs: used the original header comment from main.js

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

navyansh007 committed:
docs: modified the si and ci parameters' definition to match the sici docs

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

navyansh007 committed:
refactor: changed type of sgn from double to 8-bit signed int

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

navyansh007 committed:
refactor: implemented sincos function in main.js

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

navyansh007 commented:
Hello @gunjjoshi, thank you for your feedback. I highly appreciate your effort to mark out the suggested solutions along with the feedback. 
As mentioned in previous comments, I have implemented all the documentation changes and refactors. As requested, I have also implemented the `sincos` function in the JS implementation of the `sici` function. Feel free to let me know if any further changes are required.

gunjjoshi reviewed (dismissed):
Thanks for making the changes, @navyansh007. This PR looks good now.

Planeshifter reviewed (changes requested):
The PR is incomplete. When adding a C implementation to an existing package, the following files should be included:

  | Missing File                  | Purpose                             |
  |-------------------------------|-------------------------------------|
  | lib/native.js                 | JavaScript wrapper for native addon |
  | test/test.native.js           | Tests for native implementation     |
  | benchmark/benchmark.native.js | Benchmark for native implementation |
  
  cc @gunjjoshi

navyansh007 committed:
feat: add native addon binding for native C implementation of sici function

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

navyansh007 committed:
bench: added native benchmark for C implementation of sici

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

navyansh007 committed:
Merge branch 'stdlib-js:develop' into sici-c-implementation

navyansh007 committed:
Merge branch 'stdlib-js:develop' into sici-c-implementation

navyansh007 committed:
chore: updated the year in licenses of all modified files in sici

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

navyansh007 committed:
test: implemented test.native.js for the C implementation of sici

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

navyansh007 commented:
Hey @Planeshifter @kgryte,
I've implemented all the files requested by @Planeshifter to complete the C implementation of `sici`.

But while I was implementing and running the tests I wrote in `test/test.native.js`, the tests were failing because the tolerance `relTol * EPS * |expected|` became too small when expected values were close to 0. The actual errors were 1-4 ULPs only. Initially, I used the tolerance values as per `test/test.main.js`.

Finally, I modified it by using:
`tol = max( absoluteTol, relativeTol * abs( expected ) )`

  With tolerances:
  - Small values: max(10×EPS, 10×EPS × |expected|)
  - Medium values: max(20×EPS, 20×EPS × |expected|)
  - Large values: max(100×EPS, 100×EPS × |expected|)
  - Very large values: max(1000×EPS, 1000×EPS × |expected|)

  Is this the right approach, or should I handle the tolerance differently?

navyansh007 committed:
Merge branch 'stdlib-js:develop' into sici-c-implementation

anandkaranubc commented:
@navyansh007 You will need to refactor the tests to use [`assert/is-almost-same-value`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/assert/is-almost-same-value) instead. This also covers NaNs.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/sici/lib/polyval_gn4.js:
```diff
@@ -1,7 +1,7 @@
 * @license Apache-2.0
 *
-* Copyright (c) 2022 The Stdlib Authors.
+* Copyright (c) 2025 The Stdlib Authors.
```
The existing files will not have their copyright year changed.

navyansh007 committed:
fix(@stdlib/math/base/special/sici): improve C implementation accuracy and adjust test tolerances

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

navyansh007 committed:
chore(@stdlib/math/base/special/sici): correct copyright year in polyval files

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

navyansh007 commented:
Hey @anandkaranubc @Planeshifter 

I have fixed native tests with two changes:

1. **C implementation**: Removed simplified approximation for `x > 1e9`. Now uses the accurate auxiliary function method for all large values.

2. **Tests**: Switched to `isAlmostSameValue` with:
   - 10 ULPs for most values
   - 4100 ULPs for medium Ci(x) (needed where Ci crosses zero at x≈3.38)

At x=3.382, Ci(x)≈0.0007 has 4096 ULP difference from expected due to polynomial rounding near zero crossing.

I want to know if 4100 ULPs are acceptable for the zero-crossing region?

Planeshifter commented on lib/node_modules/@stdlib/math/base/special/sici/benchmark/c/native/benchmark.c:
```diff
@@ -0,0 +1,137 @@
+
+	print_version();
+	for ( i = 0; i < REPEATS; i++ ) {
+		printf( "# c::%s\n", NAME );
```
The benchmark label for native implementations should include `native::` to distinguish it from non-native C benchmarks. This matches the pattern in other stdlib packages like `@stdlib/math/base/special/gamma`.

```suggestion
		printf( "# c::native::%s\n", NAME );
```

navyansh007 committed:
fix(@stdlib/math/base/special/sici): refactor C implementation to reduce floating-point error

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

navyansh007 committed:
test(@stdlib/math/base/special/sici): adjust native test tolerances

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/math/base/special/sici/src/main.c:
```diff
@@ -0,0 +1,436 @@
+	double sd;
+	double cn;
+	double cd;
+	double lnax;
```
These declarations should be moved to top of function scope.

navyansh007 committed:
style(@stdlib/math/base/special/sici): move variable declarations to top of function scope

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/math/base/special/sici/src/main.c:
```diff
@@ -0,0 +1,435 @@
+	}
+
+	if ( ax > 4.0 ) {
+		double ax2;
```
Move these declarations to top of scope.

kgryte replied:
By scope, I mean top of function.

navyansh007 committed:
Merge branch 'stdlib-js:develop' into sici-c-implementation

navyansh007 committed:
refactor(math/base/special/sici): move variable declarations to top of function scope

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

navyansh007 committed:
Merge branch 'stdlib-js:develop' into sici-c-implementation
